jazz_scm-webhook/index.js

```diff
@@ -0,0 +1,366 @@
+//
+//    http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+// =========================================================================
+
+/**
+API to record SCM activity
+@author: 
+@version: 1.0
+ **/
+
+const errorHandlerModule = require("./components/error-handler.js"); //Import the error codes module.
+const responseObj = require("./components/response.js"); //Import the response module.
+const configObj = require("./components/config.js"); //Import the environment data.
+const logger = require("./components/logger.js"); //Import the logging module.
+const request = require("request");
+const moment = require("moment");
+
+module.exports.handler = (event, context, cb) => {
+
+	//Initializations
+	var errorHandler = errorHandlerModule();
+	var config = configObj(event);
+	logger.init(event, context);
+	logger.info("Webhook-events:" + JSON.stringify(event));
+	
+	if(!event || !event.body ){
+		logger.error("Unable to find the SCM activity!");
+		return cb(JSON.stringify(errorHandler.throwInputValidationError("Unable to find the SCM activity!")));
+	} 
+	var scmMap = config.SCM_MAPPINGS;
+	var scmSource, scmIdentifier = scmMap.identifier;
+
+	getScmType(scmIdentifier, event)
+	.then((result) => getScmDetails(result, event, config))
+	.then(function(result){
+		logger.info("successfuly event is updated"+JSON.stringify(result));
+		return cb(result)
+	})
+	.catch(function(error){
+		logger.info("error while updating events:"+JSON.stringify(error));
+		return cb(JSON.stringify(errorHandler.throwInternalServerError("Eventkey is null or undefined! so unable to find event name.")));
+	});
+};
+
+function getScmType(scmIdentifier, event){
+	logger.info("scmIdentifier:"+JSON.stringify(scmIdentifier));
+	return new Promise((resolve, reject) => {
+		for (var key in scmIdentifier) {
+			if(scmIdentifier.hasOwnProperty(key)) {
+				if (event.headers && event.headers.hasOwnProperty(scmIdentifier[key])){
+					scmSource = key;
+					resolve(scmSource);
+				}
+			}
+		}
+	});
+}
+
+function getScmDetails(scmSource, event, config){
+	logger.info("Inside getScmDetails:"+ scmSource)
+	return new Promise((resolve, reject) => {
+		var userName, eventKey, service, repositoryLink, servContext,
+		
+		eventBody = event.body;
+
+		if(scmSource === 'bitbucket'){
+			userName = (eventBody.actor.username) ? eventBody.actor.username : '';
+			eventKey = event.headers['X-Event-Key'];
+			service = eventBody.repository.slug;
+			repositoryLink = eventBody.repository.links.self[0].href;
+			bitbucketScmContext(eventKey, eventBody, config)
+			.then((res) => updateEvents(res, service, userName,repositoryLink, config))
```

Comment:
We should push the bitbucket code (extracting values from event and eventBody to bitbucket specific function. We should do similar thing for gitlabs.  
The resultant code can be like 

```
return new Promise((resolve, reject) => { 
      getScmContextDetails(scmSource, eventKey, eventBody, config)
      .then ((scmDetails) => updateEventsWithScmDetails(scmDetails)
      .then(function(res) { resolve(res); }
      .catch ...
}
```

where getScmContextDetails has the fork between `bitbucket` and `gitlab`
``` 
function getScmContextDetails (scmSource, eventKey, eventBody, config){
if(scmSource === 'bitbucket'){
} else if(scmSource === 'gitlab'){

}
```

Reply:
Updated the changes